parser-typechecker/src/U/Codebase/Causal/Squash.hs

```diff
@@ -0,0 +1,33 @@
+module U.Codebase.Causal.Squash (squashCausal) where
+
+import U.Codebase.Branch.Hashing qualified as Branch
+import U.Codebase.Branch.Type
+import U.Codebase.Causal (Causal (..))
+import U.Codebase.Causal.Hashing (hashCausal)
+import U.Codebase.Sqlite.Operations qualified as SqliteOps
+import U.Codebase.Sqlite.V2.HashHandle (v2HashHandle)
+import Unison.Prelude
+import Unison.Sqlite qualified as Sqlite
+
+-- Recursively discards history, resulting in a namespace tree with only single a single
+-- Causal node at every level.
+squashCausal :: CausalBranch Sqlite.Transaction -> Sqlite.Transaction (CausalBranch Sqlite.Transaction)
+squashCausal Causal {valueHash = unsquashedBranchHash, value} = do
+  runMaybeT (MaybeT (SqliteOps.tryGetSquashResult unsquashedBranchHash) >>= MaybeT . SqliteOps.loadCausalBranchByCausalHash) >>= \case
+    Just cb -> pure cb
+    Nothing -> do
+      branch@Branch {children} <- value
+      squashedChildren <- traverse squashCausal children
+      let squashedBranchHead = branch {children = squashedChildren}
+      squashedBranchHash <- Branch.hashBranch squashedBranchHead
+      let squashedCausalHash = hashCausal mempty squashedBranchHash
+      let squashedCausalBranch =
+            Causal
+              { causalHash = squashedCausalHash,
+                valueHash = squashedBranchHash,
+                parents = mempty,
+                value = pure squashedBranchHead
+              }
+      SqliteOps.saveBranch v2HashHandle squashedCausalBranch
+      SqliteOps.saveSquashResult unsquashedBranchHash squashedCausalHash
```

Comment:
I don't love that I'm manually hashing and saving the causal here; is there a better module I could move the `hashAndSaveCausalBranch` to? I can't put it in `SqliteOps` itself because we don't want any hashing stuff there.

Should I put it in `SqliteCodebase`?

Reply:
cc @mitchellwrosen

Reply:
@mitchellwrosen this function does hashing, so depends on `unison-codebase-sqlite-hashing-v2` which depends on `unison-codebase-sqlite`, so we can't put it in `unison-codebase-sqlite`.

Also in https://unisoncomputing.zulipchat.com/#narrow/stream/389079-team-arya/topic/.E2.9C.94.20Hashing.20being.20merged.20into.20codebase.20package.3F/near/366289335 you'd mentioned you brought up that we want to keep all the codebase logic separate from hashing logic.

Reply:
I think the right design for this is something like:

1. The `HashHandle` grows a function that knows how to hash a v2 causal. On inspection it looks like we don't exactly have a good type for this, so maybe we just take the hashes as separate arguments just like the existing `hashCausal`:

    ```haskell
    -- In the `HashHandle` record
    hashCausal :: Set CausalHash -> BranchHash -> CausalHash
    ```

2. Implement the `hashCausal` function concretely in the top-level `v2HashHandle`

3. Define `squashCausal` in `unison-codebase-sqlite`, passing in a `HashHandle` to provide the concrete implementation of causal hashing.

4. Pass the `v2HashHandle` explicitly at the call site(s) in `unison-parser-typechecker`.

Does that work?

Reply:
Yup that works, I'll do that 👍🏼 

I admit I don't fully understand what we gain by depending on an interface here rather than just defining all the hashing functions in one place and depending on that, but following the pattern works fine here.